### ISSUE TITLE
Not being able to change avatar

### ISSUE THREAD

**Issue by mohgh** [tinode/tindroid#85]
### Subject of the issue
In the settings, you can't change the avatar photo. You will get NotConnectedException. Also, no log will appear on server side.
### Your environment
Latest code on devel branch
#### Server-side
- [ ] web.tinode.co, api.tinode.co
- [ ] sandbox.tinode.co
- [x] Your own setup:
  * platform Linux
  * Latest
  * MySql
  * standalone

#### Client-side
- [ ] TinodeWeb/tinodejs: javascript client
  * Browser make and version.
- [x] Tindroid: Android app
  * Galaxy s7 edge
- [ ] Tinodios: iOS app
  * iOS version
  * Simulator or hardware, if hardware describe it.
- [ ] tn-cli
  * Python version
- [ ] Chatbot
  * Python version
- Version of the client, e.g. `0.15.1`
- [ ] Your own client. Describe it:
  * Transport (gRPC, websocket, long polling)
  * Programming language.
  * gRPC version, if applicable.

### Steps to reproduce
Go to settings, tap on profile photo, choose a photo from gallery. 

### Expected behaviour
Photo should be updated.

### Actual behaviour
An exception will be thrown with a toast showing you are offline.

### Server-side log
No log.

### Client-side log
2020-06-19 16:48:53.109 7552-7552/co.im.noor D/PromisedReply: REJECTING promise co.im.tinodesdk.PromisedReply@10b7961
    co.im.tinodesdk.NotConnectedException: No connection
        at co.im.tinodesdk.Tinode.send(Tinode.java:1748)
        at co.im.tinodesdk.Tinode.send(Tinode.java:1760)
        at co.im.tinodesdk.Tinode.sendWithPromise(Tinode.java:1773)
        at co.im.tinodesdk.Tinode.setMeta(Tinode.java:1582)
        at co.im.tinodesdk.Topic.setMeta(Topic.java:1119)
        at co.im.tinodesdk.Topic.setDescription(Topic.java:1137)
        at co.im.tinodesdk.Topic.setDescription(Topic.java:1149)
        at co.im.noor.UiUtils.updateAvatar(UiUtils.java:948)
        at co.im.noor.AccPersonalFragment.onActivityResult(AccPersonalFragment.java:388)
        at androidx.fragment.app.FragmentActivity.onActivityResult(FragmentActivity.java:170)
        at co.im.noor.ChatsActivity.onActivityResult(ChatsActivity.java:234)
        at android.app.Activity.dispatchActivityResult(Activity.java:7599)
        at android.app.ActivityThread.deliverResults(ActivityThread.java:4487)
        at android.app.ActivityThread.handleSendResult(ActivityThread.java:4534)
        at android.app.ActivityThread.-wrap20(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1752)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6944)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)

**Comment by or-else** [comment]
Confirmed, it's a bug. Would you like to fix it?

**Comment by mohgh** [comment]
Sure, but I couldn't find the root of it yet. Any hint what could cause this? The connection status is offline even before attempting to upload the photo, but other functionalities in the same fragment like changing the title are working fine.

**Comment by or-else** [comment]
It's a data race. The photo picker is a separate application. When photo picker is launched, Tinode app goes into background. Once in background it disconnects. When the user returns back to Tinode with the photo, it's still in disconnected state because it takes time to reconnect. 

I would keep the reference to the received photo in the activity and send it when the app reconnects instead of trying to send it on received result.

**Comment by or-else** [comment]
I just got a better idea. When the avatar is selected, show a preview screen where the user can scale and reposition the selected image within the circular view port. It will create a delay which should be enough for the app to reconnect.

**Comment by mohgh** [comment]
Regarding your first solution, How can I wait for Tinode to reconnect after resuming? Is there some kind of listener for it?

Regarding your second solution, Image cropping and scaling libraries are operating in such a way that a new activity is opened for editing the photo. So the problem will be the same or we should implement our own crop and scale tool inside a fragment in Tindroid.

**Comment by or-else** [comment]
> How can I wait for Tinode to reconnect after resuming?

You can use this listener:
https://github.com/tinode/tindroid/blob/master/app/src/main/java/co/tinode/tindroid/ChatsActivity.java#L188

> Image cropping and scaling libraries are operating in such a way that a new activity is opened for editing the photo

It's not a hard problem that requires an external library. Panning/zooming/cropping is already implemented. I would take 
https://github.com/tinode/tindroid/blob/master/app/src/main/java/co/tinode/tindroid/ImageViewFragment.java
add an optional overlay with a circular cutout in the middle, then use methods from 
https://github.com/tinode/tindroid/blob/master/app/src/main/java/co/tinode/tindroid/UiUtils.java#L419
to crop a square.

**Comment by or-else** [comment]
BTW, avatar zooming/cropping would be a very useful feature.

**Comment by mohgh** [comment]
I'm not being able to use the listener. Is this approach correct?

```
@Override
    public void onActivityResult(final int requestCode, final int resultCode, final Intent data) {
        final Activity activity = getActivity();
        if (activity == null) {
            return;
        }

        if (requestCode == UiUtils.ACTIVITY_RESULT_SELECT_PICTURE && resultCode == RESULT_OK) {
            mPhotoData = data;

            me = Cache.getTinode().getMeTopic();

            MeListener mMeTopicListener = new MeListener();
            me.setListener(mMeTopicListener);
            UiUtils.attachMeTopic(getActivity(), mMeTopicListener);
        }
    }

private class MeListener extends UiUtils.MeEventListener {

        @Override
        public void onPres(MsgServerPres pres) {
            super.onPres(pres);
        }

        @Override
        public void onOnline(boolean online) {
            Log.d("tinode profile check", online+"");
            if (mPhotoData == null){
                return;
            }
            // noinspection unchecked
            UiUtils.updateAvatar(getActivity(), me, mPhotoData);

        }
    }
```

I'm also looking into your other suggestion. Let me see how could I do it.

**Comment by or-else** [comment]
I'll implement the avatar preview feature in the next release.

**Comment by or-else** [comment]
Done in `devel`.

**Comment by or-else** [comment]
@mohgh would you like to add the same preview to TopicInfo avatar?

**Comment by sanmiguel2019** [comment]
I am able to change avatar but avatars don´t update in mobile phones. For example:

Alice uses phone 1
Bob uses phone 2
Carol uses web app

Alice changes her avatar. The new avatar immediately appears in the Carol´s web chat but in Bob´s phone Alise´s avatar does not update. Only if Bob reboots the phone, Alice´s avatar updates in the conversations list and in the conversation.

The problem persists in devel, next and in Tinode´s app on Play Store.

Avatar updates in real time are not implemented yet or something is working wrong?

**Comment by or-else** [comment]
It was a bug. Fixed: 04011f9
